monitoring/ping_test.go

```diff
@@ -0,0 +1,126 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package monitoring
+
+import (
+	"context"
+	"net"
+
+	"github.com/gravitational/satellite/agent"
+	"github.com/gravitational/satellite/agent/health"
+	"github.com/gravitational/satellite/agent/proto/agentpb"
+	serf "github.com/hashicorp/serf/client"
+	"github.com/hashicorp/serf/coordinate"
+	"gopkg.in/check.v1"
+)
+
+type PingSuite struct{}
+
+var _ = check.Suite(&PingSuite{})
+
+func (*PingSuite) TestPingChecker(c *check.C) {
+	testCases := []struct {
+		Nodes       []serf.Member
+		Coords      map[string]*coordinate.Coordinate
+		Status      agentpb.NodeStatus_Type
+		Description string
+	}{
+		{
+			Nodes: []serf.Member{
+				{
+					Name:   "member-1",
+					Status: agentpb.MemberStatus_Alive.String(),
+					Addr:   net.IPv4(127, 0, 0, 1),
+				},
+				{
+					Name:   "member-2",
+					Status: agentpb.MemberStatus_Alive.String(),
+					Addr:   net.IPv4(127, 0, 0, 2),
+				},
+			},
+			Coords: map[string]*coordinate.Coordinate{
+				"member-1": &coordinate.Coordinate{
+					Height: 0.001,
+				},
+				"member-2": &coordinate.Coordinate{
+					Height: 0.001,
+				},
+			},
+			Status:      agentpb.NodeStatus_Running,
+			Description: "Testing standard working condition with values below threshold",
+		},
+		{
+			Nodes: []serf.Member{
+				{
+					Name:   "member-1",
+					Status: agentpb.MemberStatus_Alive.String(),
+					Addr:   net.IPv4(127, 0, 0, 1),
+				},
+				{
+					Name:   "member-2",
+					Status: agentpb.MemberStatus_Alive.String(),
+					Addr:   net.IPv4(127, 0, 0, 2),
+				},
+			},
+			Coords:      map[string]*coordinate.Coordinate{},
+			Status:      agentpb.NodeStatus_Degraded,
+			Description: "Testing missing coordinates for cluster members",
+		},
+		{
+			Nodes: []serf.Member{
+				{
+					Name:   "member-1",
+					Status: agentpb.MemberStatus_Alive.String(),
+					Addr:   net.IPv4(127, 0, 0, 1),
+				},
+				{
+					Name:   "member-2",
+					Status: agentpb.MemberStatus_Failed.String(),
+					Addr:   net.IPv4(127, 0, 0, 2),
+				},
+			},
+			Coords: map[string]*coordinate.Coordinate{
+				"member-1": &coordinate.Coordinate{
+					Height: 1,
+				},
+				"member-2": &coordinate.Coordinate{
+					Height: 1,
+				},
+			},
+			Status:      agentpb.NodeStatus_Running,
+			Description: "Testing nodes with failing status",
+		},
+	}
+
+	for _, testCase := range testCases {
+		pingChecker, err := NewPingChecker(PingCheckerConfig{
+			SerfRPCAddr:    "127.0.0.1",
+			SerfMemberName: "member-1",
+			NewSerfClient: func(serf.Config) (agent.SerfClient, error) {
```

Comment:
Instead of config.SerfRPCAddr/NewSerfClient consider passing an initialized serf client to checker.

Reply:
Unless it's a huge mistake, enormous waste of resource or something that breaks other features let's postpone this improvement to a future moment when it will be needed and be done with this task, IMO.

Reply:
My point is that this indeed not needed now - i.e. complicating the design with NewSerfClient when one could be just created by the caller and passed on, no?

Reply:
ah okay. Yeah I get you're saying.

We went that way so that it would be easier to write tests... and it's done now so I won't bother changing it again. But I agree that it would have been probably easier to go the other way eventually.